content/posts/building_jnigen.md

```diff
@@ -0,0 +1,262 @@
+
+```java
+class Person {
+  String firstName;
+  @Nullable String middleName;
+  String lastName;
+  int age;
+}
+```
+
+JNIgen will wrap this into a Dart class, containing a reference to original object and accessing String fields through JNI when they are required. Further, any field of non-primitive type (including String), will return a wrapper object referring to a Java object.
+
+But when all of this is a heap of data, and if we never pass anything from this back into Java code, then it's better to convert the entire structure into a dart class at one go, and discard any references into Java.
+
+```dart
+// dart
+class Person {
+  String firstName, lastName;
+  String? middleName;
+  int age;
+  // ....
+}
+```
+
+The ~~pedantic~~  observant reader would point out that it's just deserialization in disguise.
+
+It will be certainly interesting to see whether binary-deserializing the entire thing into a byte buffer, and deserializing it back from Dart will be more efficient than populating the target POD structure field-by-field. Either way, the primary value in such conversion is ergonomics (getting a dart POD type without any effort) and not performance minutiae.
+
+## Lessons learned
+
+`jnigen` was my first project with real world scope. I learned few valuable things about software engineering in general.
+
+1. __Get the minimal end-to-end code done first, add features later:__ Later I read _"The Pragmatic Programmer"_, which calls it the "Tracer Bullets" approach. Having feedback on the code is important when developing. Few times, I made the mistake of getting lost trying to achive perfect results, despite the advice by my mentors.
+
+2. __Automate whenever possible:__ When I wrote steps including multiple commands in docs, Daco would usually suggest me to write a script for that instead. Dart streamlines this by having `tool/` directory in default project layout. This is a great advice because 1. A script is more reproducible and less error prone than a series of steps performed manually 2. Nobody reads documentation.
+
+3. __Test too much rather than test too little:__ Our assumptions turn out to be wrong all the time. Code coverage is a useful thing to have, because it helps to figure out which parts of the code are in dire need of testing. I also relied heavily on integration tests for first few iterations, which turned out to be a wrong intuition. Integration tests are more complex to write and take more time to run. So it's better to make code more unit-testable, and codify every assumption into tests.
+
+4. __Think twice before implementing:__ I made lot of decisions which were found to be suboptimal. For example, first I mapped each java package to one output Dart file, which turned out to be an uncanny valley between mapping one Dart file per class and writing all bindings to a single file. It pays to focus on plus and minus of architecture upfront.
```

Comment:
True and not true. Sometimes you need to prototype something once in order to learn what is the right way to do it.

Reply:
Changed it to `contemplate the big picture before implementing`.

I personally feel like I could've architected JNIgen better. Maybe that's normal in this field!.